### PR TITLE
⚡️ Feature fix #39: Add Parabolic SAR to enhance core trend detection

### LINKED ISSUE #39: Better core `trend` indicator for trend detection
MGM sometimes opens a trade at a peek right before a crash
Trend detection should prevent MGM to open a trade at a peek right before it crashes.

Probably caused by a detection delay.

### PR DESCRIPTION
**Pull request by andif888**
Enhances core trend detection by adding a **Parbolic SAR** Indicator to the mix. 

`.fthypt` file without sar -> [strategy_MoniGoManiHyperStrategy_no_sar.fthypt](https://github.com/Rikj000/MoniGoMani/blob/964389b0f5316fa648b025c63c1decc422ba9cef/user_data/hyperopt_results/strategy_MoniGoManiHyperStrategy_no_sar.fthypt)

`.fthypt` file with sar -> [strategy_MoniGoManiHyperStrategy_with_sar.fthypt](https://github.com/Rikj000/MoniGoMani/blob/964389b0f5316fa648b025c63c1decc422ba9cef/user_data/hyperopt_results/strategy_MoniGoManiHyperStrategy_with_sar.fthypt)

command line:
```bash
freqtrade hyperopt -s MoniGoManiHyperStrategy \
-c ./user_data/mgm-config.json \
-c ./user_data/mgm-config-private.json \
--hyperopt-loss WinRatioAndProfitRatioLoss \
--spaces all \
-e 300 \
--random-state 2 \
--timerange 20210521-
```
### without SAR
```
+--------+---------+----------+------------------+--------------+-------------------------------+-----------------+-------------+                            
|   Best |   Epoch |   Trades |    Win Draw Loss |   Avg profit |                        Profit |    Avg duration |   Objective |
|--------+---------+----------+------------------+--------------+-------------------------------+-----------------+-------------|
| * Best |   2/300 |        4 |      0    4    0 |        0.00% |                            -- | 0 days 05:44:00 |          -0 |
| * Best |   4/300 |       70 |     31   20   19 |        0.92% |        29.037 USDT    (5.81%) | 0 days 11:19:00 | -2,854.78477 |
| * Best |   9/300 |       67 |     43    2   22 |        0.80% |        24.240 USDT    (4.85%) | 0 days 08:20:00 | -3,453.67246 |                           
|   Best |  88/300 |       92 |     45    8   39 |        1.12% |        46.309 USDT    (9.26%) | 0 days 09:00:00 | -5,028.57406 |                           
 [Epoch 300 of 300 (100%)] ||   
```
### with SAR
```
+--------+---------+----------+------------------+--------------+-------------------------------+-----------------+-------------+                            
|   Best |   Epoch |   Trades |    Win Draw Loss |   Avg profit |                        Profit |    Avg duration |   Objective |
|--------+---------+----------+------------------+--------------+-------------------------------+-----------------+-------------|
| * Best |   2/300 |        5 |      0    4    1 |       -0.25% |        -0.569 USDT   (-0.11%) | 0 days 02:25:00 |           0 |
| * Best |   4/300 |       69 |     31   19   19 |        0.89% |        27.715 USDT    (5.54%) | 0 days 12:11:00 | -2,764.24830 |
| * Best |   9/300 |       52 |     32    3   17 |        1.06% |        24.748 USDT    (4.95%) | 0 days 10:37:00 | -3,380.99009 |                           
|   Best | 162/300 |      133 |     61   34   38 |        0.56% |        33.550 USDT    (6.71%) | 0 days 19:52:00 | -3,416.03917 |                           
|   Best | 279/300 |       58 |     21   28    9 |        2.10% |        54.957 USDT   (10.99%) | 0 days 20:57:00 | -4,417.39457 |                           
 [Epoch 300 of 300 (100%)] ||                                                                                       | [Time:  0:18:20, Elapsed Time: 0:18:20]
```

### CODE REVIEW

File: user_data/strategies/MasterMoniGoManiHyperStrategy.py [hunk 1]
```diff
@@ -282,13 +282,16 @@ def _populate_core_trend(self, dataframe: DataFrame, metadata: dict) -> DataFram
         # -DM (Negative Directional Indicator) = previous low - current low
         dataframe['minus_di'] = ta.MINUS_DI(dataframe, timeperiod=25)
 
+        # Parabolic SAR
+        dataframe['sar'] = ta.SAR(dataframe)
```
Comment (nuaimat) [inline user_data/strategies/MasterMoniGoManiHyperStrategy.py:286]: SAR can take these two params as well: acceleration=0.02, maximum=0.2
maybe we can HyperOpt them?

Comment (andif888) [reply user_data/strategies/MasterMoniGoManiHyperStrategy.py:286]: Indeed, would be an option. 
I am afraid, I do not understand the new code base good enough yet, to implement this option.

**Comment by raftersvk** [comment]
I have implemented the proposed solution on the v011 version and so far for me it doesn't bring a better result.
Tested with the following commands : 
freqtrade hyperopt -s MoniGoManiHyperStrategy -c ./user_data/mgm-config.json -c ./user_data/mgm-config-private.json --hyperopt-loss WinRatioAndProfitRatioLoss --spaces all -e 300 --random-state 2 --timerange 20210521- -j -2
freqtrade hyperopt-list --no-details --export-csv user_data/hyperopt_results/strategy_MGMv011_testTRENDSAR_vanilla.csv

freqtrade hyperopt -s MoniGoManiHyperStrategySAR -c ./user_data/mgm-config.json -c ./user_data/mgm-config-private.json --hyperopt-loss WinRatioAndProfitRatioLoss --spaces all -e 300 --random-state 2 --timerange 20210521- -j -2
freqtrade hyperopt-list --no-details --export-csv user_data/hyperopt_results/strategy_MGMv011_testTRENDSAR_enhanced.csv

freqtrade hyperopt -s MoniGoManiHyperStrategy -c ./user_data/mgm-config.json -c ./user_data/mgm-config-private.json --hyperopt-loss WinRatioAndProfitRatioLoss --spaces all -e 300 --random-state 20000 --timerange 20210101- -j -2
freqtrade hyperopt-list --no-details --export-csv user_data/hyperopt_results/strategy_MGMv011_testTRENDSAR_YTD_vanilla.csv

freqtrade hyperopt -s MoniGoManiHyperStrategySAR -c ./user_data/mgm-config.json -c ./user_data/mgm-config-private.json --hyperopt-loss WinRatioAndProfitRatioLoss --spaces all -e 300 --random-state 20000 --timerange 20210101- -j -2
freqtrade hyperopt-list --no-details --export-csv user_data/hyperopt_results/strategy_MGMv011_testTRENDSAR_YTD_enhanced.csv

results as zip file attached

**Comment by Rikj000** [comment]
Please run this with 1000 epochs to completion so we can truly see if there are improvements and also append the HyperOpt results table being printed out since that gives an easy overview of the results

**Comment by raftersvk** [comment]
new set of tests with 1000 epoch running... waiting for confirmation before sharing the results ;)

**Comment by Rikj000** [comment]
> so far I would say there is no real difference between the 2.
> 
> now running 2nd test on timerange 20210101-20210608

Thank you for the nice testing Rafter! :mechanical_arm: I indeed don't really see much difference either in between the 2 with the current implementation of it

**Comment by raftersvk** [comment]
2nd test on data since beginning of the year, this time it is clear that SAR combined to the actual trend gives worse results:
image

maybe there is another way this indicator should be used, but as it stands in this PR we can not integrate it

detailed files : 
https://we.tl/t-85L0MOnk5I

**Comment by andif888** [comment]
Oh dear ... 
But hands down for your accurate testing :+1: :1st_place_medal: 
I am sorry that I have caused you work.

**Comment by raftersvk** [comment]
no worries, this also why we need to test the different ideas ;)

**Comment by Rikj000** [comment]
Since we don't see improvements I'll close off this PR. But @raftersvk and @andif888, both thank you for your work and testing!

**Pull request closed without merging**